### ISSUE TITLE
Display Debug.php output in a a switchable screen in the developer admin panel

### ISSUE THREAD

**Issue by 100mountains** [100mountains/sndcore#7]
add another button at the top of the dev tab, next to debug logging on/off. call it live view on/off.

when its turned on, a screen should appear and debug.php should pipe its output to this screen internally in a secure manner. we need to ensure that its not always piping that data to the screen so have a conditional for it inside debug.php set from the database key you make for the live view on/off.

this branch previously attempted the live view with old code:

https://github.com/100mountains/sndcore/tree/copilot/fix-5

this was the last good commit:

https://github.com/100mountains/sndcore/pull/6/commits/64bc8a121a5973510d3a6fd6fc8c96f9c7816c4f

it was nearly working, the SSE endpoints were all good, but the debug was not sending messages after the initial page load. if you can think of a  better way of making the output much like 'tail -f debug.log' then go for it. this was the main idea, if you can think of a more elegant solution do it:

# Live View Debug Feature - Complete Design Documentation

## Overview
The Live View feature provides real-time streaming of debug output in the WordPress admin panel using Server-Sent Events (SSE). It transforms slow AJAX polling into smooth, continuous streaming like `tail -f debug.log` with color-coded domains and high-volume performance handling.

#### 1. Debug.php (`src/Utils/Debug.php`)
**Primary functionality**: Message capture, buffering, and rate limiting
- **Live view buffer**: In-memory array `$liveViewBuffer` stores up to 1000 entries
- **Rate limiting**: `$rateLimitBuffer` tracks timestamps to limit entries per second (default: 200/sec)
- **Domain detection**: Automatic PSR-4 namespace-based domain detection via `getDomainFromClass()` and `getDomainFromFilePath()`
- **Entry structure**: Each entry contains timestamp, message, domain, and formatted_time
- **Storage**: Uses WordPress transients for temporary storage with 5-minute expiry

**Key Methods**:
- `addToLiveView()`: Captures debug entries and adds to buffer with rate limiting
- `getLiveViewEntries()`: Retrieves entries with optional filtering and display limits
- `clearLiveView()`: Clears buffer and transients
- `isLiveViewEnabled()`: Checks if live view is enabled in configuration

#### 2. Admin.php (`src/Admin/Admin.php`)
**Primary functionality**: SSE endpoint and AJAX handlers
- **SSE streaming**: `sseStreamLiveView()` method provides real-time Server-Sent Events endpoint
- **Asset management**: Conditionally enqueues `liveview.js` when dev mode is enabled
- **Nonce handling**: Provides security nonces via `wp_localize_script()`
- **Connection management**: Handles SSE headers, buffering, and cleanup

**Key Methods**:
- `sseStreamLiveView()`: Main SSE endpoint for streaming debug entries
- `ajaxClearLiveView()`: AJAX handler for clearing the live view buffer
- `enqueueAdminAssets()`: Conditional JavaScript enqueuing

#### 3. Config.php (`src/Core/Config.php`)
**Primary functionality**: Configuration storage and defaults
- **Debug configuration**: Stores enabled state, live_view flag, rate_limit, and domain settings
- **Default values**: Provides fallback configuration when settings haven't been saved
- **Configuration structure**: 
  ```php
  '_bfp_debug' => [
      'enabled' => false,
      'live_view' => false,
      'rate_limit' => 200,
      'domains' => [...domain settings...]
  ]
  ```

#### 4. liveview.js (`assets/js/liveview.js`)
**Primary functionality**: Frontend SSE connection and UI management
- **EventSource connection**: Manages SSE connection to `/wp-admin/admin-ajax.php?action=bfp_live_view_stream`
- **UI updates**: Real-time DOM manipulation for displaying debug entries
- **Color coding**: Domain-specific color assignments for visual categorization
- **Auto-scroll**: Intelligent scrolling that respects user interaction
- **Connection management**: Auto-reconnection, error handling, and status feedback

**Key Functions**:
- `startLiveView()`: Initiates SSE connection
- `addEntryToDisplay()`: Adds new debug entries to the display
- `getDomainColor()`: Maps domains to specific colors
- `stopLiveView()`: Cleanly closes SSE connection

### 1. Debug Entry Creation
```
Application Code → Debug::log() → addToLiveView() → WordPress Transients
```

### 2. Live View Streaming
```
JavaScript SSE Request → Admin::sseStreamLiveView() → Debug::getLiveViewEntries() → JSON Response → Frontend Display
```

### 3. Configuration Management
```
Admin Settings Form → Settings::saveGlobalSettings() → Config::getDebugConfig() → Debug::isLiveViewEnabled()
```

### Rate Limiting
- **Default**: 200 entries per second
- **Implementation**: Sliding window using timestamp buffer
- **Configurable**: Can be adjusted via admin interface
- **Purpose**: Prevents UI flooding during high-volume debug scenarios

### Buffer Management
- **Buffer size**: 1000 entries maximum in memory
- **Display limit**: 500 entries shown in UI
- **Auto-cleanup**: Removes oldest entries when limits exceeded
- **Memory efficiency**: Uses array slicing for performance

### SSE Streaming
- **Connection**: Persistent EventSource connection
- **Heartbeat**: 30-second intervals to maintain connection
- **Auto-reconnect**: 3-second delay on connection loss
- **Headers**: Proper SSE headers with no-cache directives

### Authentication & Authorization
- **Capability check**: `current_user_can('manage_options')` required
- **Nonce verification**: All AJAX requests use WordPress nonces
- **Secure storage**: WordPress transients with automatic expiry

### Data Handling
- **Sanitization**: HTML escaping for all displayed content
- **Temporary storage**: 5-minute transient expiry prevents data persistence
- **No file logging**: Live view data is memory/transient only

### Modified Files
1. **src/Utils/Debug.php** - Added live view buffer system and rate limiting
2. **src/Admin/Admin.php** - Added SSE endpoint and conditional asset loading
3. **src/Core/Config.php** - Added live_view and rate_limit configuration keys
4. **assets/js/liveview.js** - NEW: Complete live view JavaScript implementation
5. **templates/dev-tools.php** - UI elements for live view toggle and display area

### JavaScript Organization
- **admin.js**: General admin functionality (unchanged)
- **liveview.js**: Dedicated live view functionality (368 lines)
- **Conditional loading**: liveview.js only loads when dev mode is enabled

### Live View Toggle
```html
<label>
    <input type="checkbox" name="_bfp_debug[live_view]" id="live_view_enabled" />
    Live View On/Off
</label>
```

### Display Area
```html
<div id="live-view-display">
    <div id="live-view-output"></div>
    <div class="live-view-controls">
        <button id="clear-live-view">Clear</button>
        <button id="auto-scroll-toggle">Auto-scroll: On</button>
    </div>
</div>
```

### CSS Styling
```css
#live-view-output {
    background: #1e1e1e;
    color: #ffffff;
    padding: 10px;
    height: 400px;
    overflow-y: auto;
    word-wrap: break-word;
    word-break: break-all;
}
```

## Domain Color Coding
- **Admin**: #4CAF50 (Green)
- **Audio**: #2196F3 (Blue)
- **Core**: #FF9800 (Orange)
- **DB**: #795548 (Brown)
- **API**: #00BCD4 (Cyan)
- **Storage**: #FFC107 (Amber)
- **UI**: #8BC34A (Light Green)
- **Utils**: #607D8B (Blue Grey)
- **WooCommerce**: #9575CD (Purple)

### SSE Connection Issues
- **Automatic reconnection**: 3-second delay with retry logic
- **Status feedback**: Visual indicators for connection state
- **Graceful degradation**: Falls back to connection loss message

### Rate Limit Handling
- **Silent dropping**: Excess entries are simply not added to buffer
- **Timestamp tracking**: 1-second sliding window for rate calculation
- **Memory management**: Auto-cleanup of expired rate limit entries

### Buffer Overflow
- **Automatic trimming**: Removes oldest entries when buffer exceeds 1000
- **Display limiting**: UI shows maximum 500 entries
- **Performance protection**: Prevents memory issues during high-volume logging

### Debug Settings
```php
'_bfp_debug' => [
    'enabled' => false,        // Master debug toggle
    'live_view' => false,      // Live view streaming toggle
    'rate_limit' => 200,       // Max entries per second
    'domains' => [             // Individual domain toggles
        'admin' => false,
        'core' => false,
        // ... additional domains
    ]
]
```

### WordPress Integration
- **Settings API**: Uses WordPress options for persistence
- **Transients**: Temporary storage with automatic cleanup
- **Hooks**: Integrates with WordPress admin_init and admin_enqueue_scripts
- **Nonces**: WordPress security for all AJAX operations

### Memory Management
- **Buffer limits**: Hard caps on buffer size and display entries
- **Transient expiry**: 5-minute automatic cleanup
- **Rate limiting**: Prevents excessive memory usage

### Network Efficiency
- **SSE streaming**: Single persistent connection vs polling
- **JSON compression**: Efficient data serialization
- **Selective loading**: JavaScript only loads when needed

### UI Performance
- **DOM limiting**: Maximum 500 displayed entries
- **Efficient scrolling**: Smart auto-scroll detection
- **CSS optimization**: Hardware-accelerated scrolling
- **Text wrapping**: Prevents horizontal layout issues

### WordPress Hooks
- `admin_enqueue_scripts`: Asset loading
- `wp_ajax_bfp_live_view_stream`: SSE endpoint
- `wp_ajax_bfp_live_view`: Clear functionality

### PSR-4 Namespace Detection
- **Automatic domain detection**: Based on class namespace structure
- **File path fallback**: When class information unavailable
- **Manual override**: Domain can be specified explicitly

### WooCommerce Integration
- **Conditional loading**: Only when WooCommerce is active
- **Domain detection**: Automatic detection of WC-related debug entries
- **Product context**: Debug entries can include product information

This design provides a professional, high-performance debug streaming solution that integrates seamlessly with the existing WordPress plugin architecture while maintaining security and performance standards.
